appsdk/configurable.go

```diff
@@ -0,0 +1,176 @@
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+
+package appsdk
+
+import (
+	"strconv"
+	"strings"
+
+	"github.com/edgexfoundry/go-mod-core-contracts/models"
+
+	"github.com/edgexfoundry/app-functions-sdk-go/pkg/transforms"
+)
+
+// AppFunctionsSDKConfigurable contains the helper functions that return the function pointers for building the configurable function pipeline.
+// They transform the parameters map from the Pipeline configuration in to the actual actual parameters required by the function.
+type AppFunctionsSDKConfigurable struct {
+	Sdk *AppFunctionsSDK
+}
+
+// DeviceNameFilter - Specify the devices of interest to  filter for data coming from certain sensors.
+// The Filter by Device transform looks at the Event in the message and looks at the devices of interest list,
+// provided by this function, and filters out those messages whose Event is for devices not on the
+// devices of interest.
+// This function will return an error and stop the pipeline if a non-edgex
+// event is received or if no data is recieved.
+// For example, data generated by a motor does not get passed to functions only interested in data from a thermostat.
+// This function is a configuration function and returns a function pointer.
+func (dynamic AppFunctionsSDKConfigurable) DeviceNameFilter(parameters map[string]string) AppFunction {
+	deviceNames, ok := parameters["DeviceNames"]
+	if !ok {
+		dynamic.Sdk.LoggingClient.Error("Could not find deviceNames")
+	}
+
+	transform := transforms.Filter{
+		FilterValues: strings.Fields(deviceNames),
+	}
+	return transform.FilterByDeviceName
+}
+
+// ValueDescriptorFilter - Specify the value descriptors of interest to filter for data from certain types of IoT objects,
+// such as temperatures, motion, and so forth, that may come from an array of sensors or devices. The Filter by Value Descriptor assesses
+// the data in each Event and Reading, and removes readings that have a value descriptor that is not in the list of
+// value descriptors of interest for the application.
+// This function will return an error and stop the pipeline if a non-edgex
+// event is received or if no data is recieved.
+// For example, pressure reading data does not go to functions only interested in motion data.
+// This function is a configuration function and returns a function pointer.
+func (dynamic AppFunctionsSDKConfigurable) ValueDescriptorFilter(parameters map[string]string) AppFunction {
+	valueDescriptors, ok := parameters["ValueDescriptors"]
+	if !ok {
+		dynamic.Sdk.LoggingClient.Error("Could not find ValueDescriptors")
+	}
+	transform := transforms.Filter{
+		FilterValues: strings.Fields(valueDescriptors),
```

Comment:
As we found out, the names can have spaces, so the ValueDescriptors need have a none white space separator .